Backport py3.11 asyncio's taskgroups.

smurfix opened the pull request:
This backports 3.11's TaskGroup class to uasyncio.

I don't care much about the `except *` stuff, not in µPy context anyway, but sane error recovery is crucial and taskgroups make this job a whole lot easier, not to mention *much* less bug prone. (Writing from a lot of Trio and anyio experience here.)

Bottom line: I refuse to write async code without using taskgroups. So here you are.

~TODO: Write a couple of tests.~

Closes #8508.

smurfix committed:
extmod/uasyncio: Backport py3.11 asyncio's taskgroups.

This patch adds a mostly-compatible backport of CPython 3.11's
`TaskGroup` class to uasyncio.

Also, there is a new `run_server` method in uasyncio/stream.py which
supports task groups and doesn't run in the background (no need).

TODO: Write some more tests.

Closes #8508.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
*Sigh*. It's not my commit message that's failing the test!

dpgeorge commented:
Wow, thanks for this!  It's very nice to see that it can be implemented in pure Python.  I don't know anything about TaskGroup so will need to learn how it works.

dpgeorge commented:
> It's not my commit message that's failing the test!

You're right.  That CI check will need fixing so it allows "Revert" commit messages.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+# Adapted with permission from the EdgeDB project.
+# Adapted for MicroPython
```
We need to be careful with licensing.  If this code was copied from CPython, that's possibly a show stopper.  The rest of uasyncio is MIT licensed and has original authorship specifically for MicroPython.

smurfix replied:
The original is Apache licensed. That's sufficiently compatible with MIT (afaik) that I don't think we even need to ask for permission, though I intend to do that shortly anyway.

dpgeorge commented:
>> It's not my commit message that's failing the test!
>
> You're right. That CI check will need fixing so it allows "Revert" commit messages.

Actually the CI check is working OK, it's only supposed to check the commit messages that are unique to the PR.

If you rebase (and force push) on latest master it will hopefully get the CI green.

smurfix commented:
Pushed. Writing tests, found a significant bug (related to cancellation of course). Investigating.

smurfix committed:
extmod/uasyncio: Task needs to be hashable.

The C implementation (_uasyncio.Task) was not.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
Huh. I have no idea where those test failures are coming from.

smurfix commented:
> It looks like the bytecode emitter allows 2 args but the native emitter only allows 1.

Well, yeah, I saw that line, but what has that got to do with implementing `__hash__` and/or how do I fix that? I didn't touch no emitters … so either I'm doing something really dumb here, or I found a bug which I'm not qualified to fix, esp. since the code works on my system(s).

dlech commented:
Assuming you are using the unix port, to reproduce locally, try `make -C ports/unix test_full`. 

There are a number of tests that are disabled when emit=native because of `raise_varargs` already, `basics/try_reraise.py`, `basics/try_reraise2` and `basics/try_finally_return2.py` to name a few (search for `raise_varargs` in `tests/run-tests.py`).

I suspect you have written some Python code similar to one of these that causes the same error.

peterhinch commented:
I'm running this on a Pyboard 1.1 - TaskGroups are a very nice feature.

Something Google seems loath to divulge is how to access return values from the members of a group (assuming they terminate normally).

smurfix commented:
@peterhinch Glad that my code is of use. I'll try to fix up the CI errors shortly; battling with Covid aftereffects, so if somebody else wants to fix up this PR, feel free.

You need to return any values explicitly, e.g. by setting an object attribute to it or queuing the value or whatever.

peterhinch commented:
Sorry you're unwell - I hope you feel better soon.
> Writing tests, found a significant bug (related to cancellation of course). Investigating.

Is this still outstanding or is the code ready for review?

I very much hope this is implemented: in many applications it's a big improvement over `gather`.

smurfix commented:
> Is this still outstanding or is the code ready for review?

No, that's done. The workaround I implemented admittedly isn't particularly clean, but it gets the job done and doesn't alter non-taskgroup code (as that might introduce bugs or slowdowns).

peterhinch commented:
@dpgeorge @jimmo  This is to advocate for `TaskGroup` with a real world example where they fit the bill perfectly.

Consider a communication link between two peers using an unreliable medium such as WiFi. This may be near the limit of range and the link can fail in a variety of ways. The simplest way to recover is the "belt and braces" approach: if a peer encounters an unrecoverable error, it takes the link down for a period long enough for the other peer also to suffer an unrecoverable error. That way both peers start from a "power up" state.

A `TaskGroup` might then comprise:
 * Task trying to keep the WiFi available through brief outages.
 * Task waiting on received messages.
 * Task pinging the other peer and checking a response.
 * Intermittently a task is added to the group. It sends a message, awaits an ack, and terminates.

If any of these experience an error that cannot be handled locally, the exception is passed up to be handled by the task that created the group. When this occurs, every task in the group terminates in an orderly way, running cleanup code in `finally` clauses or via context managers.

The `TaskGroup` is ideal for this - and in my testing works fine :) I rest my case...

smurfix commented:
@peterhinch Exactly. Unstructured tasks (like asyncio's baseline tasks) mean that you have to keep track of what's still running and what might have to be cancelled and/or restarted on your own. Code doing that tends to contain a heap of bugs you can't test for, and basically doesn't scale.

In contrast, when a taskgroup's async context manager has ended you *know* that there's no dangling tasks, unprocessed `finally` clauses, unreported exceptions, or similar nonsense around; you can proceed with confidence. When (not if)  the framework ensures that whole classes of errors can't happen, you no longer have to test for them, or even think about them.

The latter point is much more powerful than you'd assume at first glance.

smurfix commented:
OK, so I have no idea why this patch could possibly cause this error. Open to suggestions.

peterhinch commented:
My approach would be to run `tests/extmod/uasyncio_heaplock.py` on the PC and revert changes to see where the problem lies. Starting with changes to `stream.py`.

dpgeorge commented:
> OK, so I have no idea why this patch could possibly cause this error. Open to suggestions.

The issue is that the native emitter doesn't support `async with`, which is now used in `uasyncio/stream.py`.

I think it's important for uasyncio to still work with the native emitter.  Since the native emitter won't support `async with` in the near future the best way to deal with this would be to convert the use of `async with` to other code which doesn't explicitly use the `async with` construct.  Eg use a try/finally instead and await the calls to `__aenter__` and `__aexit__`.

smurfix commented:
Ugh. Thanks, I'll do that.

smurfix commented:
OK, so now I get another one which I don't have any idea how to fix (and which doesn't happen on my system …).
```
micropython: ../../py/emitnative.c:2817: emit_native_raise_varargs: Assertion `n_args == 1' failed.
```

dpgeorge commented on extmod/uasyncio/stream.py:
```diff
@@ -170,6 +171,54 @@ async def start_server(cb, host, port, backlog=5):
+            await _run_server(tg, s, cb)
+        except BaseException as exc:
+            if not await _tg.__aexit__(type(exc), exc, None):
+                raise
```
use `raise exc` and it should work with the native emitter

smurfix commented:
*Whew*.

I'd appreciate a merge, assuming you have no further comments.

dpgeorge commented:
This will need some tests, ideally to test all new functions, classes and methods introduced by this PR.  They can be stand-alone (in `tests/extmod/`) and/or multi-tests (in `tests/multi_net`).  You can look at the existing uasyncio tests for templates and ideas.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+# Adapted with permission from the EdgeDB project.
```
This will need a MicroPython copyright and license header, like the other uasyncio source files.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+# Adapted for MicroPython
+
+
+__all__ = ["TaskGroup"]
```
This is not supported by MicroPython so best to remove it.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+from . import event
+from . import task
+
+DEBUG = False
```
If you do `_DEBUG = const(False)` then this will be completely removed by the compiler when false.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+        self._base_error = None
+        self._on_completed = None
+
+    def __repr__(self):
```
Is it necessary to have a complicated repr like this?

smurfix replied:
I'll reduce it to something simpler when _DEBUG is False.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+    def __repr__(self):
+        info = [""]
+        if self._tasks:
+            info.append(f"tasks={len(self._tasks)}")
```
Please don't use f-strings, because they are not available on all ports/boards.  Use `.format()` instead.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+    def __init__(self):
+        self._entered = False
+        self._exiting = False
+        self._aborting = False
```
Is it possible to combine these 3 into one `self._state` variable?

smurfix replied:
Sure. Will do that.

dpgeorge commented on extmod/uasyncio/taskgroup.py:
```diff
@@ -0,0 +1,244 @@
+
+        k = [None]
+        t = self._loop.create_task(self._run_task(k, coro))
+        k[0] = t  # sigh
```
Why does it need to use a list with one element?

smurfix replied:
Because you need to pass the newly created task to `_run_task`, which you can't otherwise do without some uasyncio surgery, which I wanted to avoid.

smurfix commented:
One remaining problem: the C `Task` class in `moduasyncio.c` requires a `__hash__` method. Having no experience with the intricacies of implementing C-level extensions to MicroPython I have no idea what to do. (The hash should simply be the object's address, or the low bytes thereof.)

dhalbert commented:
A semantics q: how is this implementation of TaskGroup affected by the lack of exception groups? My understanding is that TaskGroup was held up in CPython until exception groups were worked out and then implemented? Thjanks.

smurfix commented:
> how is this implementation of TaskGroup affected by the lack of exception groups?

Well, `class ExceptionGroup` already is in `uasyncio.core`, albeit without multiple inheritance (an `ExceptionGroup` should be both an `Exception` and a `BaseExceptionGroup`) and without helpers to filter it.

What's lacking is said filtering, but that's not exactly trivial. It's also not very important for embedded systems IMHO. The important part is the ability to say "The part of my program that does X needs to be stopped / restarted due to an error. Do this cleanly and without requiring any boilerplate code". A `TaskGroup` can do this without problems even if the `ExceptionGroup` class is just a container.

dhalbert commented:
Thanks, so the TaskGroup implementation does not directly miss the absence of  `except *`, etc. I agree with you that it's not very important; I was wondering if you had to make changes iin the TaskGroup code due to the missing Python 3.11 features.

smurfix commented:
If somebody could help out with the `_hash__` issue (https://github.com/micropython/micropython/pull/8791#issuecomment-1434556915) I'd be exceedingly happy. ;-)

dpgeorge commented:
> If somebody could help out with the `_hash__` issue

It should be enough to add the following line to the `MP_DEFINE_CONST_OBJ_TYPE` for `task_type` (before the line with `task_attr`):
```
unary_op, mp_generic_unary_op,
```

smurfix commented:
No idea what to do about the coverage problem. IMHO this is ready to merge.

smurfix commented:
codecov again …

Anyway, the last two commits address improved exception handling, in that we want to see the original exception(s) instead of the cancellation of the task that started them. I hope not to uncover more of these.

jimmo commented:
@smurfix This is looking really good! Thanks for all your work on it.

Question about the tests... We provide .exp files for the asyncio tests mostly because the tests need to run on older Python versions that don't have asyncio. So adding a .exp file for uasyncio_taskgroup is the right thing to do, but the output should match what CPython (3.11 in this case) would emit.

I tried this -- for test_taskgroup_04, MicroPython prints `ZeroDivision` but CPython raises it.

@dpgeorge Perhaps we should consider adding a .pyexp extension to distinguish between "this requires latest CPython" and "this is known different behavior / cannot run on CPython".

jimmo commented:
(And if this is a known/expected difference, e.g. due to something like #10818, then a comment in the test code would be required).

smurfix commented:
> I tried this -- for test_taskgroup_04, MicroPython prints ZeroDivision but CPython raises it.

Not any more.

smurfix commented:
To explain further: CPython's asyncio considers a task to be runnable as soon as the timeout expires and won't cancel it until the next time it calls an awaitable. uasyncio on the other hand processes them sequentially and thus cancels the task immediately.

Also, currently a single exception is passed through instead of being wrapped in an `ExceptionGroup` because that's the most common case. MicroPython doesn't support `except*` syntax – and it doesn't have exceptiongroup split-up methods either. 

I consider that to be an implementation detail, thus I modified the affected tests so that they work both ways.

I'll add (a condensed version of) this to the "differences to Python 3.11" file – as soon as somebody creates that :-P

smurfix commented:
Another corner case showed up in the project I need all of this for.

peterhinch commented:
@smurfix What is the status of this PR?

smurfix commented:
Works perfectly for me …

I'll do a rebase shortly.

smurfix commented:
Any idea why the taskgroup test fails with `--emit=native` would be appreciated.

smurfix committed:
extmod/uasyncio/taskgroup: Rewrite state machine.

Tasks can be cancelled before they have an opportunity to run.
Handling this situation is non-trivial in MicroPython's context.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
I have pushed an update to the taskgroup state machine, which unfortunately crashes the test.

This version crashes in one of the tests.

```
[…]
--- 08
Not handled: divide by zero
D True

Program received signal SIGSEGV, Segmentation fault.
mp_execute_bytecode (code_state=code_state@entry=0x7ffff7ad7530, inject_exc=inject_exc@entry=0x0)
    at ../../py/vm.c:1372
1372	            MP_STATE_VM(cur_exception) = nlr.ret_val;
(gdb) whe
#0  mp_execute_bytecode (code_state=code_state@entry=0x7ffff7ad7530, inject_exc=inject_exc@entry=0x0)
    at ../../py/vm.c:1372
#1  0x000055555559581e in mp_obj_gen_resume (self_in=self_in@entry=0x7ffff7ad7520, 
    send_value=send_value@entry=0x6, throw_value=throw_value@entry=0x0, 
    ret_val=ret_val@entry=0x7fffffffd2d8) at ../../py/objgenerator.c:219
#2  0x000055555558c6a4 in mp_resume (self_in=0x7ffff7ad7520, send_value=0x6, 
    throw_value=throw_value@entry=0x0, ret_val=ret_val@entry=0x7fffffffd2d8) at ../../py/runtime.c:1403
#3  0x00005555555a780d in mp_execute_bytecode (code_state=code_state@entry=0x7ffff7acb450, 
    inject_exc=inject_exc@entry=0x0) at ../../py/vm.c:1201
#4  0x000055555559581e in mp_obj_gen_resume (self_in=0x7ffff7acb440, send_value=<optimized out>, 
    throw_value=throw_value@entry=0x0, ret_val=ret_val@entry=0x7fffffffd3a8)
    at ../../py/objgenerator.c:219
#5  0x00005555555958ab in gen_resume_and_raise (self_in=<optimized out>, send_value=<optimized out>, 
    throw_value=throw_value@entry=0x0, raise_stop_iteration=raise_stop_iteration@entry=true)
    at ../../py/objgenerator.c:268
#6  0x0000555555595945 in gen_instance_send (self_in=<optimized out>, send_value=<optimized out>)
    at ../../py/objgenerator.c:295
#7  0x0000555555594e32 in fun_builtin_2_call (self_in=0x55555562c360 <gen_instance_send_obj>, 
    n_args=<optimized out>, n_kw=<optimized out>, args=0x7ffff7acb7f0) at ../../py/objfun.c:80
#8  0x000055555558c196 in mp_call_function_n_kw (fun_in=0x55555562c360 <gen_instance_send_obj>, 
    n_args=2, n_kw=0, args=0x7ffff7acb7f0) at ../../py/runtime.c:707
#9  0x000055555558c2df in mp_call_method_n_kw (n_args=<optimized out>, n_kw=<optimized out>, 
    args=args@entry=0x7ffff7acb7e8) at ../../py/runtime.c:723
#10 0x00005555555a66a9 in mp_execute_bytecode (code_state=code_state@entry=0x7ffff7acb7c0, 
    inject_exc=inject_exc@entry=0x0) at ../../py/vm.c:1042
#11 0x000055555559509b in fun_bc_call (self_in=0x7ffff7aca000, n_args=1, n_kw=0, args=0x7fffffffd6a0)
    at ../../py/objfun.c:273
[…]
#24 0x00005555555e7d3c in main (argc=2, argv=0x7fffffffdc08) at main.c:476

```

 I'd appreciate help with debugging this; frankly I have no idea why this should happen.

smurfix committed:
extmod/uasyncio/core: Handle colliding sleep_ms calls.

If two tasks try to start sleeping at the same time,
the SingletonGenerator will be in use, thus we need
to allocate a new temporary handler on the heap.

As the singleton isn't that single any more, it gets renamed.

Found by taskgroup test #12.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix committed:
extmod/asyncio/taskgroup.py: State machine update.

This fixes some edge cases with cancellation.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix committed:
tests/extmod/asyncio_taskgroup: Improve test output.

still failing, though

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
Update: found a workaround, still chasing a (different, and OF COURSE cancellation related) bug …

smurfix committed:
extmod/asyncio/core: Implement BaseExceptionGroup.split().

Rudimentary but works.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix committed:
extmod/asyncio/taskgroup: Colliding cancel vs. exception.

If there's an exception (or more) *and* a cancellation in a taskgroup,
the cancellation must be ignored.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
Added a workaround for the previous crash.

I have no freakin'  idea what's triggering the thread_exc1 error.

smurfix committed:
extmod/asyncio/taskgroup: Exiting vs. aborting taskgroups.

This mirrors Python change 594c369.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
I really do hope that this is now finally ready for merging.

andrewleech commented on extmod/asyncio/stream.py:
```diff
@@ -17,7 +17,8 @@ async def __aenter__(self):
     async def __aexit__(self, exc_type, exc, tb):
-        await self.close()
+        self.s.close()
+        pass
```
Why was await removed here? Also pass shouldn't be needed

smurfix replied:
The `await` never made any sense here since the `close` method isn't async and in fact never was. It just works by coincidence.

You're right about the `pass` of course, I'll fix that up.

projectgus replied:
Very minor, but marking this as unresolved because the `pass` is still here.

andrewleech commented on extmod/asyncio/core.py:
```diff
@@ -51,9 +70,10 @@ def __next__(self):
+# Try not to allocate a SleepHandler on the heap if possible
+def sleep_ms(t, sgen=SleepHandler()):
+    if sgen.state is not None:  # the static one is busy
+        sgen = SleepHandler()
```
Could you give some details on when / how the singleton is already in use?
It seems like it'll be confusing that sometimes sleep will allocate and sometimes it won't.

smurfix replied:
Of course. Simply start two tasks and let both of them sleep at the same time.

I don't think it's confusing, this is entirely internal to `sleep_ms`.

andrewleech replied:
I very regularly have a large number of tasks with each of them sleeping for different lengths of time... I haven't noticed any issue. 

What problem / symptoms do you see?

smurfix replied:
I initially found this bug via test 12 in `tests/extmod/asyncio_taskgroup.py`, according to my notes, though I don't remember details.

This test no longer errors out, but adding a `print` statement to the `if` clause does show that the problem is still triggered by the test, and that it does cause the sleep time of *some* other task to be modified and/or ignored.

The non-augmented code depends on the scheduler to iterate the "singleton" object that's yielded by `sleep_ms` immediately, i.e.. before any other task that also happens to be scheduled calls `sleep_ms`. This obviously holds most of the time, but not always.

andrewleech replied:
> The original code depends on the scheduler to iterate the "singleton" object that's yielded by `sleep_ms` immediately, i.e.. before any other task which also happens to be scheduled calls `sleep_ms`. This consition seems to hold most of the time, but not always.

Ah, this is quite interesting. I had wondered how the Singleton could ever work to provide multiple different sleep function usages.... I can see how multiple tasks started back to back (like in a new taskgroup setup) would be more likely to trigger this situation of multiple definitions before the task loop makes it back around to the start to process the sleeps.

andrewleech commented on extmod/asyncio/core.py:
```diff
@@ -222,6 +242,9 @@ def run_until_complete(main_task=None):
                 _exc_context["exception"] = exc
                 _exc_context["future"] = t
                 Loop.call_exception_handler(_exc_context)
+                # XXX if we do await it later,
```
I'm not sure what this comment means, can you elaborate?

smurfix replied:
`it` in this context is the task. I'll change the comment.

This is a core bug which, before taskgroups, nobody seems to have hit yet.

andrewleech replied:
Oh, await'ing a task that's already been await / run. Yes that's known to be not supported I guess, I hadn't thought of it as an issue as such :-)

andrewleech reviewed (commented):
TaskGroup certainly looks interesting, a useful syntax for sure so thanks for putting this together.

I might be missing something... it feels a lot like `gather` except that it's a context manager and automatically cancels other tasks if one raises an exception?

Is there more to it than this?

smurfix commented on extmod/asyncio/stream.py:
```diff
@@ -170,6 +171,47 @@ async def start_server(cb, host, port, backlog=5):
 
+# Helper task to run a TCP stream server.
+# Callbacks (i.e. connection handlers) may run in a different taskgroup.
+async def run_server(cb, host, port, backlog=5, taskgroup=None):
```
cpython doesn't have a taskgroup-aware way to run a server, it simply starts a new task for each client (for some value of "simply", its code is rather complex for historical reasons related to their protocol/transport split – in hindsight, that was a rather bad design decision).

That process doesn't play well with taskgroups. A taskgroup *guarantees* that when you leave it, all of its resources, subtasks and whatnot are done with. As soon as you start a non-taskgrouped task from within that, all those nice guarantees go out the window. You can't work around this problem without introducing a race condition. 

Yes this is a micropython extension and thus something that we ordinarily don't want to do, but CPython simply doesn't have an equivalent.

smurfix replied:
… which reminds me that the whole taskgroup thing needs documentation. Duh. Working on it.

andrewleech replied:
So on cpython, if you run start _server in a TaskGroup it'll run, but not get cleaned up if the TaskGroup exits?

Do you know if this has raised as an issue for cpython?

smurfix replied:
I have no idea. I plan to investigate that (and submit a PR if necessary).

The "problem" is that most likely the issue hasn't come up yet because people who use taskgroups tend to do that with anyio or even trio. Both have their own wrappers to address this problem, which unfortunately are not a good fit for micropython.

smurfix replied:
No I didn't yet, no free time to shepherd a feature like that, esp. since taskgroups still are an under-utilized (and poorly understood, among old-time asyncio adherents) feature.

I'll drop it for now and move the code to `examples/` until we find a better home for it.

smurfix commented:
> Is there more to it than this?

Well, yes. `gather` is static, you start a fixed list of tasks and then wait for them. A taskgroup is dynamic, you can start new tasks whenever you need to, provided you have the taskgroup object.

Also the fact that it is a context manager is much more powerful than appears at first glance because you *know* that when the group's context manager exits, all the tasks in it are *done*. Thus you can restart part or all of your system, in case of an error or a config change or whatever, without adding "oops I have subtasks, I need to remember to cancel them" exception recovery code all over the place (and not just as a wrapper around your eventual call to `gather`!). We all know how well-tested *that* kind of code tends to be.

A good introduction is https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/

smurfix committed:
extmod/asyncio/taskgroup: NO-OP cancelling an inactive taskgroup.

Cancelling a taskgroup that's no longer active should not do anything.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix committed:
docs/library/asyncio: Add a section on taskgroups.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

andrewleech commented:
> Also the fact that it is a context manager is much more powerful

Oh don't get me wrong I definitely support this new syntax, it looks very useful and I've just recently written annoying / messy code to run/cancel multiple concurrent tasks that this would be perfect for!

I was actually thinking more from an implementation perspective, it feels to me you could possibly reuse most of gather code, but putting all tasks into a list that can be appended to. With  return_exceptions=False it'll exit at first exception and then the new code would need to cancel the rest of the tasks from the list.

... I haven't actually reviewed the code to this level to know if this suggestion should be feasible.

smurfix commented:
> ... I haven't actually reviewed the code to this level to know if this suggestion should be feasible.

The devil is in the detail. Like that `RuntimeError("can't gather")` which gets raised when `gather` can't set its hook.

On the other hand, the `task.data` hook might actually be a workable replacement of the wrapper procedure I'm currently using because, surprise, a taskgroup always creates its own subtasks instead of getting them passed in from somewhere else. I'll add this to my list of possible improvements, but IMHO this question shouldn't block merging.

smurfix commented:
Yes I know this is one of >300 open PRs … nevertheless, I'd appreciate feedback WRT this patch's mergeability.

smurfix committed:
extmod/asyncio/taskgroup: do the more common test first

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix committed:
extmod/asyncio/taskgroup: handle self-cancellation

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix committed:
extmod/asyncio/taskgroup: print errors to stderr

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
Ugh. Don't ask me what this CI error means …

projectgus reviewed (commented):
Coming in late to this. I haven't used Taskgroups in CPython but I can see the utility - this looks really useful. Thanks for all the work and persistence.

> Ugh. Don't ask me what this CI error means …

```
+Traceback (most recent call last):
+  File "<stdin>", line 13, in <module>
+AttributeError: 'module' object has no attribute 'ExceptionGroup'
+

FAILURE /home/runner/work/micropython/micropython/tests/results/extmod_asyncio_taskgroup.py
```

`webassembly` port has its own version of the asyncio module under `ports/webassembly/asyncio`. I think you've got two options here:

1) Add `extmod/asyncio/taskgroup.py` to `ports/webassembly/variants/manifest.py` and `ports/webassembly/asyncio/__init__.py`
2) Or update the test to skip if the asyncio module doesn't contain this name

dpgeorge commented on extmod/asyncio/stream.py:
```diff
@@ -13,11 +14,19 @@ def __init__(self, s, e={}):
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.close()
```
Close is not async, so how does this work?

smurfix replied:
It doesn't … another merge oversight I'm afraid.

smurfix commented on extmod/asyncio/core.py:
```diff
@@ -229,6 +249,9 @@ def run_until_complete(main_task=None):
                 Loop.call_exception_handler(_exc_context)
+                # XXX if we do await the task later,
+                # leaving t.data as None will cause a fault.
+                t.data = exc
```
That's a merge artefact. Fixing.

smurfix committed:
docs/asyncio: Fix capitalization error.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
     ignore this exception.  Cleanup code may be run by trapping it, or via
     ``try ... finally``.
 
+class Taskgroup
```
Right. Fixing.

smurfix committed:
asyncio: move `run_server` to examples.

`run_server` is not in CPython. We like to avoid that.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented on tests/run-tests.py:
```diff
@@ -622,6 +622,9 @@ def run_tests(pyb, tests, args, result_dir, num_threads=1):
         skip_tests.add("extmod/ssl_poll.py")
+    elif args.emit == "native":
+        skip_tests.add("extmod/asyncio_heaplock.py")
+        skip_tests.add("thread/thread_exc1.py")
```
Correct. See https://github.com/micropython/micropython/pull/8791#issuecomment-1164802135

smurfix committed:
webassembly/asyncio: Add support for taskgroups.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix committed:
ports/webassembly: Add exception groups to its asyncio module.

Signed-off-by: Matthias Urlichs <matthias@urlichs.de>

smurfix commented:
Updated to current master.

The CPython discussion on what to do about the `Taskgroup.cancel` method has stalled for now (and mainly centered on the *name* of the feature, as opposed to how it's supposed to work). That doesn't surprise me because the semantics of cancellation are somewhat murky in asyncio, but as cancelling all tasks in a taskgroup is something reasonable to do …

>  `+mpy-cross: ../py/emitnative.c:3010: emit_native_raise_varargs: Assertion `n_args == 1' failed.`

Umm, that used not to happen AFAIR?

>  It should be that the scheduler always iterates the singleton immediately

Well, it appears not to. In any case, by its nature async code is async: nothing prevents me from calling a bare `asyncio.sleep` and then doing something else before using the result in an `await`. Thus we shouldn't depend on that whether or not it's supposed to apply in the standard case.

```python
    async def not_too_fast(delay, proc):
        dly = asyncio.sleep(delay)
        res = await proc()
        await dly
        return res
```
In MicroPython this is a nicely concise way of saying "give me the result of `proc` in exactly two seconds, except when `proc` takes longer than that`. (This doesn't work the same way on CPython, but not by design AFAICT.)

m-u-xyz commented:
A patch that adds `taskgroup.cancel()`, with basically the same semantics as this PR's, has been merged into CPython.

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
 
+class TaskGroup
+---------------
+
```
At the moment the link at the top of this page is to CPython 3.8's asyncio reference, which doesn't have TaskGroups. Suggest starting this section with a short explanation of this being a port of the CPython 3.11 feature, and a link to https://docs.python.org/3.11/library/asyncio-task.html#task-groups

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+class TaskGroup
+---------------
+
+See Nathaniel J. Smith's `essay on Structured Concurrency
+<https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/>`_
+for an introduction why you should use taskgroups instead of starting
+"naked" tasks.
+
+.. note::
+    His "nursery" objects are called "taskgroup" in asyncio; the
+    equivalent of a "go statement" is `Loop.create_task`.
```
This is interesting context and motivation, but I'm a bit torn on whether this is suitable for the MicroPython asyncio reference or not. (i.e. the equivalent CPython docs provide a single sentence summary of what TaskGroups do and that feels more appropriate for reference documentation.)

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+
+.. class:: TaskGroup()
+
+    This object is an async context managed holding a group of tasks.
+    Tasks can be added to the group using `TaskGroup.create_task`.
```
```suggestion
    An asynchronous context manager holding a group of tasks. Tasks can be
    added to the group using `TaskGroup.create_task`. All tasks are awaited when
    the context manager exits.
```

(Originally this was just a suggestion to fix the typo, but changed it to copying the CPython doc summary exactly!)

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+    This method is equivalent to cancelling the task responsible for the
+    body of the taskgroup, *if* that is what the task is currently doing.
+
+.. exception:: Cancelled
```
This exception doesn't seem to exist in CPython, it shares `asyncio.CancelledError` for this case - doesn't it?

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+    returned in the first list.
+
+    MicroPython does not support CPython 3.11's syntax for filtering handling
+    exception groups.
```
```suggestion
    MicroPython does not support CPython 3.11's ``except*`` syntax for handling
    exception groups without calling this function explicitly.
```

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -215,22 +285,41 @@ TCP stream connections
 
     This is a coroutine.
 
+.. function:: run_server(callback, host, port, backlog=5, taskgroup=None)
```
I have two concerns here:

1. This function doesn't exist in CPython 3.11 asyncio, and we're trying not to add any more MicroPython-only names in CPython-compatible modules..
2. This function doesn't appear to exist in MicroPython asyncio module either, at least in this version of the PR it's in the `examples/run_server.py` file. Ergonomically this is a bit rough for someone who wants to use it.

Is the comment above about how `start_server` doesn't work well with TaskGroups exclusive to MicroPython? If so, is there any way that we can change MicroPython's implementation of `start_server` to be closer to CPython's in this regard?

(Otherwise, if `start_server` also doesn't work well with TaskGroups on CPython then a good path forward might be to add `run_server` as a micropython-lib library that can be separately `mip install`-ed, and mention it in the docs for `start_server` here.)

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -326,6 +421,8 @@ Event Loop
 
     Create a task from the given *coro* and return the new `Task` object.
 
+    You should not call this function when you're using taskgroups.
```
```suggestion
    .. warning: Due to MicroPython internal limitations, this function can't be used at the same time as TaskGroups.
```

projectgus commented on extmod/asyncio/taskgroup.py:
```diff
@@ -0,0 +1,267 @@
+# Adapted with permission from the EdgeDB project.
```
Should we include EdgeDB's copyright header here as well?

projectgus commented on extmod/asyncio/taskgroup.py:
```diff
@@ -0,0 +1,267 @@
+            self._errors.append(exc)
+
+        if self._errors:
+            # Exceptions are heavy objects that can have object
+            # cycles (bad for GC); let's not keep a reference to
```
This is a bit nitpicky, but MicroPython does mark & collect GC not reference counting so I don't think object cycles are so bad, really.

The comment is right that they're heavy objects. Although I think in most cases the task group object will fall out of scope at roughly the same time as `__aexit__` anyway, right? In which case the optimisation here won't make a difference to when all of these objects are GCed or how long the GC takes (because if the task group is already unreachable then the GC won't ever examine `self._errors`).

So it might be worth either removing this small optimisation or removing the comment about object cycles, although it's not a big deal either way.

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+    This method is equivalent to cancelling the task responsible for the
+    body of the taskgroup, *if* that is what the task is currently doing.
+
+.. exception:: Cancelled
+
+    This exception is raised in a task whose taskgroup is being cancelled.
+
+    This is a subclass of ``BaseException``; it should never be caught.
```
```suggestion
.. exception:: CancelledError

   The operation has been cancelled.

   This exception can be caught to perform custom operations when asyncio Tasks are cancelled. In almost all situations the exception must be re-raised.

    This is a subclass of ``BaseException``.
```

Somehow `CancelledError` wasn't documented in MicroPython until now! The above suggestion is cribbed from the CPython docs, but I think their advice to catch and re-raise is more useful than blanket "should never be caught".

(There's a follow-up task for someone to change all the occurences of `asyncio.CancelledError` in this doc to be reference links, but you don't have to do that chore in this PR!)

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+
+    This method is equivalent to cancelling the task responsible for the
+    body of the taskgroup, *if* that is what the task is currently doing.
+
```
```suggestion
    .. note:: This is a MicroPython-only method.
```

(Similarly, if there's anywhere else here that a function or behaviour is different to CPython then please add an explanatory note.)

~~I noticed that CPython's solution to this has been added to the docs in recent versions:
https://docs.python.org/3.14/library/asyncio-task.html#terminating-a-task-group~~

~~Curious if you think that's a viable alternative to having a `TaskGroup.cancel()` method?~~ EDIT: nevermind!

projectgus replied:
> Curious if you think that's a viable alternative to having a `TaskGroup.cancel()` method?

Oh, I missed your comment that CPython is getting this function! In which case maybe can say that here instead (that it's a CPython 3.15...? feature?)

projectgus commented on ports/webassembly/asyncio/__init__.py:
```diff
@@ -5,5 +5,6 @@
+from .taskgroup import TaskGroup
 
-__version__ = (3, 0, 0)
+__version__ = (3, 0, 1)
```
```suggestion
__version__ = (3, 1, 0)
```

Nitpicky, but adding a whole new feature is at least worth a minor version bump!

projectgus commented on tests/extmod/asyncio_taskgroup.py:
```diff
@@ -0,0 +1,841 @@
+# Test Lock class
```
```suggestion
# Test TaskGroups
```

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+    If multiple subtasks raise exceptions in parallel, it's unclear which
+    of them should be propagated. Thus an `ExceptionGroup` exception
+    collects them and is raised instead.
+
```
```suggestion
    .. note:: In CPython ExceptionGroup is a built-in exception, but in MicroPython it is
       part of the asyncio module.
        
       MicroPython does not support CPython 3.11's ``except*`` syntax for handling
       exception groups automatically.
```

projectgus commented on docs/library/asyncio.rst:
```diff
@@ -112,6 +112,76 @@ class Task
+    callable that returns ``True`` if the exception passed to it should be
+    returned in the first list.
+
+    MicroPython does not support CPython 3.11's syntax for filtering handling
+    exception groups.
+
```
```suggestion
```

Suggest removing this from here and covering it in the class documentation above.

projectgus commented on tests/extmod/asyncio_taskgroup.py:
```diff
@@ -0,0 +1,841 @@
+
+
+async def test_taskgroup_08():
+    if sys.implementation.name != "micropython":
+        print("Not handled: divide by zero")
```
Looks like some of these tests can run on CPython (as the top of the file checks if ExceptionGroup is a built-in), but not all of them (hence the .exp file and these checks)?

Not asking you to necessarily change anything, but how much work would it be to have tests which run on CPython & MicroPython instead?

projectgus reviewed (changes requested):
Hi @smurfix,

Sorry I'm coming in very late to this! Thanks for keeping the PR open and up to date for so long.

I have some comments but they're mostly docs and CPython differences. The implementation looks very clean and well commented, and the test coverage is comprehensive which is great.

Apologies if anything I've commented on here has previously discussed, I skimmed the comment history but I haven't read every comment - sorry.

Certainly I'm not asking for any major additional work, don't want perfect to become the enemy of good here as this looks really useful to have in MicroPython! Are you still interested in working on it?
